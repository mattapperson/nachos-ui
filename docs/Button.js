import React from 'react'
import { View } from 'react-native'

import Button from '../src/Button'
import H4 from '../src/typography/H4'

// WEBPACK (ButtonExample)
const ButtonExample = () => {
  const btnStyle = { margin: 3 }
  return (
    <View style={{ flex: 1, flexDirection: 'column' }}>
      <H4>Example:</H4>
      <Button type='success' style={btnStyle}>Success</Button>
      <Button type='danger' style={btnStyle}>Danger</Button>
      <Button style={btnStyle}>Primary</Button>

      <Button kind='squared' type='success' style={btnStyle}>
        Success
      </Button>
      <Button kind='squared' type='danger' style={btnStyle}>
        Danger
      </Button>
      <Button
        kind='squared'
        iconName='md-cloud-download'
        style={btnStyle}
      >
        Primary
      </Button>

      <H4 align='center'>Disabled style</H4>
      <Button type='success' disabled style={btnStyle}>
        Success
      </Button>
      <Button kind='squared' type='danger' disabled style={btnStyle}>
        Danger
      </Button>
    </View>
  )
}
// WEBPACK

const description = `
| Prop | Type | Default | Description |
|------|------|---------|-------------|
| \`activeOpacity\` | \`number\` | \`0.8\` | Determines the opacity of the Button when touch is active |
| \`disabled\` | \`bool\` | \`false\` | Determines whether the Button should be disabled |
| \`children\` | \`string, nodes\` | \`void\` | Add string of text or additional React components  |
| \`style\` | \`object, array\` | \`void\` | Style the Button compoment |
| \`textStyle\` | \`object, array\` | \`void\` | Style the Button's Text compoment |
| \`disabledStyle\` | \`object, array\` | \`{ opacity: 0.3 }\` | Disable style of Button |
| \`disabledTextStyle\` | \`object, array\` | \`void\` | Disable style of Button's text |
| \`type\` | \`enum\` | \`primary\` | One of: \`primary, danger, success, naked\` |
| \`kind\` | \`enum\` | \`rounded\` | One of: \`rounded, squared\` |
| \`onPress\` | \`func\` | \`void\` | Called when the touch is released |
| \`onPressIn\` | \`func\` | \`void\` | Function |
| \`onPressOut\` | \`func\` | \`void\` | Function |
| \`onLongPress\` | \`func\` | \`void\` | Function |
| \`iconName\` | \`string\` | \`void\` | Specifies an icon name. It uses [Ionicons](http://ionicframework.com/docs/v2/ionicons/)  |
| \`iconSize\` | \`number\` | \`void\` | Specifies an icon size |
| \`iconColor\` | \`string\` | \`void\` | Specifies an icon color |
| \`iconActiveColor\` | \`string\` | \`void\` | Specifies a selected icon color |
| \`iconPosition\` | \`enum\` | \`'right'\` | One of: \`left, right\` |
| \`uppercase\` | \`bool\` | \`'true'\` | Determines whether text should be uppercased  |
| \`theme\` | \`object\` | \`defaultTheme\` | An object describing a style of the component. See more in Theme section.  |
`

ButtonExample.styleguide = {
  ...ButtonExample.styleguide,
  title: 'Button',
  description,
}

export default ButtonExample
