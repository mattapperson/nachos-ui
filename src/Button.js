import React from 'react'
import PropTypes from 'prop-types'
import { View, Text, TouchableOpacity } from 'react-native'
import Icon from 'react-native-vector-icons/dist/Ionicons'
import themeManager from './themeManager'

const defaultTheme = {
  BUTTON_FONT_COLOR: '#fff',
  BUTTON_ROUNDED_RADIUS: 25,
  BUTTON_ROUNDED_HEIGHT: 50,
  BUTTON_ROUNDED_FONT_SIZE: 12,
  BUTTON_SQUARED_HEIGHT: 50,
  BUTTON_SQUARED_FONT_SIZE: 12,
  BUTTON_STATE_SUCCESS: '#94de45',
  BUTTON_STATE_DANGER: '#ff9c00',
  BUTTON_STATE_PRIMARY: '#2f8cff',
  BUTTON_ICON_SIZE: 25,
  BUTTON_ICON_COLOR: '#fff',
  BUTTON_ACTIVE_ICON_COLOR: 'rgba(0, 0, 0, 0.5)',
}

const Button = (props) => {
  const {
    activeOpacity,
    disabled,
    children,
    style,
    textStyle,
    disabledStyle,
    disabledTextStyle,
    type,
    kind,
    uppercase,
    iconName,
    iconSize,
    iconColor,
    iconActiveColor,
    iconPosition,
    onPress,
    onPressIn,
    onPressOut,
    onLongPress,
    // NOTE: injected by a Switcher
    selected,
    onChange,
    value,
  } = props

  const theme = props.theme || themeManager.getStyle('Button')
  const baseStyle = defaultStyle(theme)

  // NOTE: function onChange is injected by the Switcher component
  const switcherProp = onChange &&
    {
      onPress: () => {},
      onPressOut: onChange.bind(null, value),
    }

  const touchableProps = {
    onPress,
    onPressIn,
    onPressOut,
    onLongPress,
  }

  const btnStyles = [
    baseStyle.baseBtn,
    baseStyle.kind[kind].btn,
    baseStyle.states[type],
    style,
  ]

  const textStyles = [
    baseStyle.baseText,
    baseStyle.kind[kind].text,
    textStyle,
    disabled ? disabledTextStyle : {},
  ]

  let leftIcon
  let rightIcon
  if (iconName) {
    const icon = (
      <Icon
        name={iconName}
        size={iconSize || theme.BUTTON_ICON_SIZE}
        color={
          selected
            ? iconActiveColor || theme.BUTTON_ACTIVE_ICON_COLOR
            : iconColor || theme.BUTTON_ICON_COLOR
        }
      />
    )
    leftIcon = iconPosition === 'left' && icon
    rightIcon = iconPosition === 'right' && icon
  }

  let content
  if (children) {
    content = (
      <Text style={textStyles}>
        {uppercase ? children.toUpperCase() : children}
      </Text>
    )
  }

  return (
    <View
      style={[baseStyle.container, disabled ? disabledStyle : {}]}
    >
      <TouchableOpacity
        {...touchableProps}
        {...switcherProp}
        disabled={disabled}
        style={btnStyles}
        activeOpacity={activeOpacity}
        accessibilityTraits='button'
        accessibilityComponentType='button'
      >
        <View style={baseStyle.innerContainer}>
          {leftIcon}
          {content}
          {rightIcon}
        </View>
      </TouchableOpacity>
    </View>
  )
}

Button.defaultStyle = (theme = defaultTheme) => {
  return {
    container: { flex: 1 },
    baseBtn: {
      justifyContent: 'center',
      padding: 10,
      paddingHorizontal: 20,
    },
    baseText: {
      alignSelf: 'center',
      color: theme.BUTTON_FONT_COLOR,
      paddingHorizontal: 8,
      fontWeight: '600',
    },
    innerContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
    },
    kind: {
      rounded: {
        btn: {
          borderRadius: theme.BUTTON_ROUNDED_RADIUS,
          height: theme.BUTTON_ROUNDED_HEIGHT,
        },
        text: {
          fontSize: theme.BUTTON_ROUNDED_FONT_SIZE,
          fontWeight: '600',
        },
      },
      squared: {
        btn: { height: theme.BUTTON_SQUARED_HEIGHT },
        text: { fontSize: theme.BUTTON_SQUARED_FONT_SIZE },
      },
    },
    states: {
      success: { backgroundColor: theme.BUTTON_STATE_SUCCESS },
      danger: { backgroundColor: theme.BUTTON_STATE_DANGER },
      primary: { backgroundColor: theme.BUTTON_STATE_PRIMARY },
      naked: { backgroundColor: 'transparent' },
    },
  }
}

Button.propTypes = {
  activeOpacity: PropTypes.number,
  disabled: PropTypes.bool,
  children: PropTypes.any,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  textStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  disabledStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  disabledTextStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  type: PropTypes.oneOf(['primary', 'danger', 'success', 'naked']),
  kind: PropTypes.oneOf(['rounded', 'squared']),
  onPress: PropTypes.func,
  onPressIn: PropTypes.func,
  onPressOut: PropTypes.func,
  onLongPress: PropTypes.func,
  iconName: PropTypes.string,
  iconSize: PropTypes.number,
  iconColor: PropTypes.string,
  iconActiveColor: PropTypes.string,
  iconPosition: PropTypes.oneOf(['left', 'right']),
  uppercase: PropTypes.bool,
  theme: PropTypes.object,
  // NOTE: injected by a Switcher
  onChange: PropTypes.func,
  selected: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
}

Button.defaultProps = {
  activeOpacity: 0.8,
  disabled: false,
  disabledStyle: { opacity: 0.3 },
  type: 'primary',
  kind: 'rounded',
  onPress: () => {},
  onPressIn: () => {},
  onPressOut: () => {},
  onLongPress: () => {},
  iconPosition: 'right',
  uppercase: true,
}

export default Button
