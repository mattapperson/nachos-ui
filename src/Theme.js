import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
