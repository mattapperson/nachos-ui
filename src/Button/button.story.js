import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import React from "react";
import Button from "./index";
import { View } from "react-native";

const btnStyle = { margin: 3 };

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Button type="success" iconName="md-cloud-download" style={btnStyle}>
        Success
      </Button>
      <Button type="danger" style={btnStyle}>
        Danger
      </Button>
      <Button style={btnStyle}>Primary</Button>

      <Button type="success" disabled style={btnStyle}>
        Disabled Primary
      </Button>
    </View>
  );
};

storiesOf("Button", module).add("default", withInfo()(Example));

export default Example;
