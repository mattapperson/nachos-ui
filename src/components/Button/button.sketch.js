import React from "react";
import Button from "./index";

const btnStyle = { margin: 3 };

export const components = [
  {
    Component: () => (
      <Button type="success" iconName="person" style={btnStyle}>
        Success....
      </Button>
    ),
    name: "Button/Success"
  },
  {
    Component: () => (
      <Button type="danger" style={btnStyle}>
        Danger
      </Button>
    ),
    name: "Button/Danger"
  },
  {
    Component: () => <Button style={btnStyle}>Primary</Button>,
    name: "Button/Primary"
  },
  {
    Component: () => (
      <Button type="success" disabled style={btnStyle}>
        Disabled Primary
      </Button>
    ),
    name: "Button/Disabled"
  }
];
