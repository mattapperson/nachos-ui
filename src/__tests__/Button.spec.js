import "react-native";
import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import Button from "../Button";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Button>Submit</Button>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render disabled", () => {
  const wrapper = shallow(
    <Provider>
      <Button disabled>Submit</Button>
    </Provider>
  );
  expect(wrapper.find("TouchableOpacity").prop("disabled")).toBe(true);
});

it("should call onPress callback", () => {
  const callback = jest.fn();
  const wrapper = shallow(
    <Provider>
      <Button onPress={callback}>Submit</Button>
    </Provider>
  );
  wrapper.find("TouchableOpacity").simulate("press");
  expect(callback).toBeCalled();
});
