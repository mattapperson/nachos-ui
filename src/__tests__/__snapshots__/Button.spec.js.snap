// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  style={
    Object {
      "WebkitFlex": 1,
      "WebkitFlexGrow": "1 !important",
      "WebkitFlexShrink": "1 !important",
      "flex": 1,
      "flexGrow": "1 !important",
      "flexShrink": "1 !important",
      "msFlex": 1,
      "msFlexNegative": "1 !important",
      "msFlexPositive": "1 !important",
    }
  }
>
  <div
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    disabled={false}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    style={
      Object {
        "WebkitBoxPack": "center",
        "WebkitJustifyContent": "center",
        "backgroundColor": "rgba(47,140,255,1)",
        "borderBottomLeftRadius": "25px",
        "borderBottomRightRadius": "25px",
        "borderTopLeftRadius": "25px",
        "borderTopRightRadius": "25px",
        "height": "50px",
        "justifyContent": "center",
        "msFlexPack": "center",
        "paddingBottom": "10px",
        "paddingLeft": "20px",
        "paddingRight": "20px",
        "paddingTop": "10px",
      }
    }
    tabIndex="0"
  >
    <div
      className="rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
      style={
        Object {
          "WebkitAlignItems": "center",
          "WebkitBoxAlign": "center",
          "WebkitBoxDirection": "normal",
          "WebkitBoxOrient": "horizontal",
          "WebkitBoxPack": "center",
          "WebkitFlexDirection": "row",
          "WebkitJustifyContent": "center",
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "msFlexAlign": "center",
          "msFlexDirection": "row",
          "msFlexPack": "center",
        }
      }
    >
      <div
        className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-1471scf rn-fontFamily-14xgk7a rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-lineHeight-t9a87b rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingBottom-1mdbw0j rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
        dir="auto"
        style={
          Object {
            "WebkitAlignSelf": "center",
            "alignSelf": "center",
            "color": "rgba(255,255,255,1)",
            "fontSize": "12px",
            "fontWeight": "600",
            "msFlexItemAlign": "center",
            "paddingLeft": "8px",
            "paddingRight": "8px",
          }
        }
      >
        SUBMIT
      </div>
    </div>
  </div>
</div>
`;
